cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp

```diff
@@ -380,35 +456,783 @@ table_with_metadata hybrid_scan_reader_impl::materialize_payload_columns(
+  out_columns.reserve(_output_buffers.size());
+
+  // Copy number of total input row groups and number of surviving row groups from predicate
+  // pushdown.
+  out_metadata.num_input_row_groups = _file_itm_data.num_input_row_groups;
+  // Copy the number surviving row groups from each predicate pushdown only if the filter has
+  // value.
+  if (_expr_conv.get_converted_expr().has_value()) {
+    out_metadata.num_row_groups_after_stats_filter =
+      _file_itm_data.surviving_row_groups.after_stats_filter;
+    out_metadata.num_row_groups_after_bloom_filter =
+      _file_itm_data.surviving_row_groups.after_bloom_filter;
+  }
+
+  // no work to do (this can happen on the first pass if we have no rows to read)
+  if (!has_more_work()) {
+    // Finalize output
+    return finalize_output(read_columns_mode, out_metadata, out_columns, row_mask);
+  }
+
+  auto& pass            = *_pass_itm_data;
+  auto& subpass         = *pass.subpass;
+  auto const& read_info = subpass.output_chunk_read_info[subpass.current_output_chunk];
+
+  // Allocate memory buffers for the output columns.
+  allocate_columns(read_mode::READ_ALL, read_info.skip_rows, read_info.num_rows);
+
+  // Parse data into the output buffers.
+  decode_page_data(read_info.skip_rows, read_info.num_rows);
+
+  // Create the final output cudf columns.
+  for (size_t i = 0; i < _output_buffers.size(); ++i) {
+    auto metadata           = _reader_column_schema.has_value()
+                                ? std::make_optional<reader_column_schema>((*_reader_column_schema)[i])
+                                : std::nullopt;
+    auto const& schema      = _extended_metadata->get_schema(_output_column_schemas[i]);
+    auto const logical_type = schema.logical_type.value_or(LogicalType{});
+    // FIXED_LEN_BYTE_ARRAY never read as string.
+    // TODO: if we ever decide that the default reader behavior is to treat unannotated BINARY
+    // as binary and not strings, this test needs to change.
+    if (schema.type == Type::FIXED_LEN_BYTE_ARRAY and logical_type.type != LogicalType::DECIMAL) {
+      metadata = std::make_optional<reader_column_schema>();
+      metadata->set_convert_binary_to_strings(false);
+      metadata->set_type_length(schema.type_length);
+    }
+    // Only construct `out_metadata` if `_output_metadata` has not been cached.
+    if (!_output_metadata) {
+      cudf::io::column_name_info& col_name = out_metadata.schema_info[i];
+      out_columns.emplace_back(make_column(_output_buffers[i], &col_name, metadata, _stream));
+    } else {
+      out_columns.emplace_back(make_column(_output_buffers[i], nullptr, metadata, _stream));
+    }
+  }
+
+  // Add empty columns if needed. Filter output columns based on filter.
+  return finalize_output(read_columns_mode, out_metadata, out_columns, row_mask);
+}
+
+void hybrid_scan_reader_impl::decode_page_data(size_t skip_rows, size_t num_rows)
+{
+  auto& pass    = *_pass_itm_data;
```

Comment:
Identical to the base reader impl except for a one liner change. See comment below